README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
<h1 align="center" style="font-size: 3rem;">
algo-CLI
</h1>

Print common algorithms to the command line


# Index

* [Installation](#installation)
* [Description](#description)
* [How to use](#how-to-use)
  - [Options](#options)
  - [Supported Languages](#supported-languages)
  - [Supported Algorithms](#supported-algorithms)
* [Examples](#examples)
* [Available Themes](#available-themes)
* [Contributions](#contributions)
* [Credits](#credits)


# Installation

Use pip:

    $ pip install algocli

Manual installation:

    $ git clone https://github.com/emanuel2718/algocli.git
    $ cd algocli
    $ python setup.py install


# Description

**algocli** is a command-line tool that lets users print common algorithms directly into the terminal. Why open
a browser search through countless articles about how to do [insertionsort in python](#insertionsort-python) when you can just type `algocli python insertionsort` in the terminal.

    $ algocli [INPUT ...] [OPTIONS]
    
    
![bubblesortPython](https://user-images.githubusercontent.com/55965894/110212969-cb68e780-7e52-11eb-8f34-1b02a11b4c56.png)

&nbsp; 

# How to use

&nbsp; 

## Options:

    -h, --help              show this help message and exit
    -v, --version           displays the current version of algcli
    --list-colors           print list of available colorschemes
    --list-lang             print list of supported languages
    --list-algo             print list of supported algorithms
    
    -c [COLORSCHEME], --color [COLORSCHEME]
                            colorized output
    

## Supported Languages:

Correct language query (shown on the left) must be given for **algocli** to understand

    actionscript         Actionscript
    ada                  Ada
    algol68              ALGOL68
    applescript          Applescript
    autohotkey           Autohotkey
    awk                  AWK
    c                    C
    cpp                  C++
    csharp               C#
    d                    D
    delphi               Delphi
    fsharp               F#
    eiffel               Eiffel
    fortran              Fortran
    go                   Go
    haskell              Haskell
    objc                 Objective-C
    java                 Java
    javascript           Javascript
    lua                  Lua
    matlab               Matlab
    ocaml                Ocaml
    pascal               Pascal
    perl                 Perl
    php                  PHP
    powershell           PowerShell
    python               Python
    ruby                 Ruby
    rust                 Rust
    scala                Scala
    swift                Swift
    
## Supported Algorithms:

Correct language query (shown on the left) must be given for **algocli** to understand

    avltrees             AVL Trees
    b64                  Decode Base64 data
    beadsort             Bead Sort algorithm
    binarysearch         Binary Search algorithm
    bogosort             Bogo Sort algorithm
    bubblesort           Bubble Sort algorithm
    caesarcipher         Caesar Cipher
    cocktailsort         Cocktail Sort algorithm
    combsort             Comb Sort algorithm
    countingsort         Counting Sort algorithm
    cyclesort            Cycle Sort algorithm
    damm                 Damm algorithm
    dijkstra             Dijkstra algorithm
    e                    Calculate the value of e
    eulermethod          Euler method
    evolutionary         Evolutionary algorithm
    factorial            Calculate factorials
    factorions           Calculate factorions
    fft                  Fast Fourier Transforms
    fib                  Fibonacci Sequence
    fibnstep             Fibonacci N-step Number Sequence
    fileexists           Check if a given file exists or not
    fizzbuzz             FizzBuzz
    floydwarshall        Floy Warshall algorithm
    gnomesort            Gnome Sort algorithm
    hammingnumbers       Hamming numbers
    heapsort             Heap Sort algorithm
    huffman              Huffman coding
    insertionsort        Insertion Sort algorithm
    isaac                ISAAC Cipher
    knapsack             Knapsack Problem 0-1
    knapsackbound        Knapsack Problem Bounded
    knapsackcont         Knapsack Problem Continous
    knapsackunbound      Knapsack Problem Unbounded
    kolakoski            Kolakoski Sequence
    mandelbrot           Mandelbrot Set
    mazegen              Maze Generation
    mazesolve            Maze Solving
    md4                  How to use MD4
    md5                  How to use MD5
    md5imp               MD5 Algorithm implementation
    mergesort            Merge Sort algorithm
    nqueen               N-Queens Problem
    pancakesort          Pancake Sort algorithm
    patiencesort         Patience Sort algorithm
    permutationsort      Permutation Sort algorithm
    quickselect          Quickselect Algorithm
    quicksort            Quick Sort algorithm
    radixsort            Radix Sort algorithm
    recaman              Recaman Sequence
    regex                Simple Regular Expressions
    rot13                Rot-13 Algorithm
    rsa                  RSA code
    selectionsort        Selection Sort algorithm
    sexyprime            Sexy primes
    sha1                 SHA-1 Algorithm
    sha256               SHA-256 Algorithm
    shellsort            Shell Sort algorithm
    sieve                Sieve of Eratosthenes Algorithm
    sleepsort            Sleep Sort algorithm
    stoogesort           Stooge Sort algorithm
    strandsort           Strand Sort algorithm
    subcipher            Substitution Cipher
    toposort             Topological Sort Algorithm

&nbsp; 

# Examples

**NOTE**: The order of the options do not matter, but for the sake of simplicity all the examples will be shown with the algorithm first followed by the language. The following are equivalent:
    
    $ algocli radixsort cpp
    $ algocli cpp radixsort
    
&nbsp; 

#### List of supported languages
    
    $ algocli --list-lang

#### List of supported algorithms

    $ algocli --list-algo

#### Insertion Sort with Python *without* color

    $ algocli insertionsort python

#### Insertion Sort with Python *with* color

    $ algocli insertionsort python -c

#### Insertion Sort with Python *with* material colorscheme

    $ algocli insertionsort python -c material
    
#### Radix Sort with C++

    $ algocli radixsort cpp
    
#### Fibonacci Sequence calculation with Java

    $ algocli fib java

&nbsp; 

# Available Themes

[List of available Themes](https://github.com/emanuel2718/algocli/blob/master/THEMES.md)

&nbsp; 

    
# Contributions

*insert contributions message here*

&nbsp; 

# Credits

This project couldn't have been possible without [Rosetta Code](https://www.rosettacode.org/wiki/Rosetta_Code), which is a wonderful resource for any programmer looking to learn about how to do different things in almost any programming language in existence.

All credits go to Rosetta Code and all the contributors of the site.





CHANGELOG.md
# Changelog


## [1.1] - 2021-04-07

### Added

- First algoCLI release


algocli/util.py
'''
Utility data structures
Written by Emanuel Ramirez (emanuel2718@gmail.com)
'''


def get_available_colors():
    ''' Returns the a list of available pygments.styles colorschemes

    NOTE:   Can't be a hardcoded list of selected colorschemes because
            some colorschemes are added by external plugins.
            Plus some colors are not available in all Operating systems.

    '''
    from pygments.styles import get_all_styles
    return list(get_all_styles())


'''
    ALGORITHMS configuration

    key: user input algorithm flag
    value[0]:  Rosseta Code algorithms format
    value[1]:  Formal algorithm name to be displayed as output title
'''
ALGORITHMS = {
    'astar': ['A*_search_algorithm', 'A* Search algorithm'],
    'avltrees': ['AVL_tree', 'AVL Trees'],
    'b64': ['Base64_decode_data', 'Decode Base64 data'],
    'beadsort': ['Sorting_algorithms/Bead_sort', 'Bead Sort algorithm'],
    'binarysearch': ['Binary_search', 'Binary Search algorithm'],
    'bogosort': ['Sorting_algorithms/Bogosort', 'Bogo Sort algorithm'],
    'bubblesort': ['Sorting_algorithms/Bubble_sort', 'Bubble Sort algorithm'],
    'caesarcipher': ['Caesar_cipher', 'Caesar Cipher'],
    'cocktailsort': ['Sorting_algorithms/Cocktail_sort', 'Cocktail Sort algorithm'],
    'combsort': ['Sorting_algorithms/Comb_sort', 'Comb Sort algorithm'],
    'countingsort': ['Sorting_algorithms/Counting_sort', 'Counting Sort algorithm'],
    'cyclesort': ['Sorting_algorithms/Cycle_sort', 'Cycle Sort algorithm'],
    'damm': ['Damm_algorithm', 'Damm algorithm'],
    'dijkstra': ['Dijkstra%27s_algorithm', 'Dijkstra algorithm'],
    'e': ['Calculating_the_value_of_e', 'Calculate the value of e'],
    'eulermethod': ['Euler_method', 'Euler method'],
    'evolutionary': ['Evolutionary_algorithm', 'Evolutionary algorithm'],
    'factorial': ['Factorial', 'Calculate factorials'],
    'factorions': ['Factorions', 'Calculate factorions'],
    'fft': ['Fast_Fourier_transform', 'Fast Fourier Transforms'],
    'fib': ['Fibonacci_sequence', 'Fibonacci Sequence'],
    'fibnstep': ['Fibonacci_n-step_number_sequences', 'Fibonacci N-step Number Sequence'],
    'fileexists': ['Check_that_file_exists', 'Check if a given file exists or not'],
    'fizzbuzz': ['FizzBuzz', 'FizzBuzz'],
    'floydwarshall': ['Floyd-Warshall_algorithm', 'Floy Warshall algorithm'],
    'gnomesort': ['Sorting_algorithms/Gnome_sort', 'Gnome Sort algorithm'],
    'hammingnumbers': ['Hamming_numbers', 'Hamming numbers'],
    'heapsort': ['Sorting_algorithms/Heapsort', 'Heap Sort algorithm'],
    'helloworld': ['Hello_world/Text', 'Print Hello world in the given language'],
    'huffman': ['Huffman_coding', 'Huffman coding'],
    'insertionsort': ['Sorting_algorithms/Insertion_sort', 'Insertion Sort algorithm'],
    'isaac': ['The_ISAAC_Cipher', 'ISAAC Cipher'],
    'knapsack': ['Knapsack_problem/0-1', 'Knapsack Problem 0-1'],
    'knapsackbound': ['Knapsack_problem/Bounded', 'Knapsack Problem Bounded'],
    'knapsackunbound': ['Knapsack_problem/Unbounded', 'Knapsack Problem Unbounded'],
    'kolakoski': ['Kolakoski_sequence', 'Kolakoski Sequence'],
    'mandelbrot': ['Mandelbrot_set', 'Mandelbrot Set'],
    'mazegen': ['Maze_generation', 'Maze Generation'],
    'mazesolve': ['Maze_solving', 'Maze Solving'],
    'md4': ['MD4', 'How to use MD4'],
    'md5': ['MD5', 'How to use MD5'],
    'md5imp': ['MD5/Implementation', 'MD5 Algorithm implementation'],
    'mergesort': ['Sorting_algorithms/Merge_sort', 'Merge Sort algorithm'],
    'nqueen': ['N-queens_problem', 'N-Queens Problem'],
    'pancakesort': ['Sorting_algorithms/Pancake_sort', 'Pancake Sort algorithm'],
    'patiencesort': ['Sorting_algorithms/Patience_sort', 'Patience Sort algorithm'],
    'permutationsort': ['Sorting_algorithms/Permutation_sort', 'Permutation Sort algorithm'],
    'quickselect': ['Quickselect_algorithm', 'Quickselect Algorithm'],
    'quicksort': ['Sorting_algorithms/Quicksort', 'Quick Sort algorithm'],
    'radixsort': ['Sorting_algorithms/Radix_sort', 'Radix Sort algorithm'],
    'randgen': ['Random_number_generator_(included)', 'Random number generator'],
    'recaman': ['Recaman%27s_sequence', 'Recaman Sequence'],
    'regex': ['Regular_expressions', 'Simple Regular Expressions'],
    'rot13': ['Rot-13', 'Rot-13 Algorithm'],
    'rsa': ['RSA_code', 'RSA code'],
    'selectionsort': ['Sorting_algorithms/Selection_sort', 'Selection Sort algorithm'],
    'sexyprime': ['Sexy_primes', 'Sexy primes'],
    'sha1': ['SHA-1', 'SHA-1 Algorithm'],
    'sha256': ['SHA-256', 'SHA-256 Algorithm'],
    'shellsort': ['Sorting_algorithms/Shell_sort', 'Shell Sort algorithm'],
    'sieve': ['Sieve_of_Eratosthenes', 'Sieve of Eratosthenes Algorithm'],
    'sleepsort': ['Sorting_algorithms/Sleep_sort', 'Sleep Sort algorithm'],
    'stoogesort': ['Sorting_algorithms/Stooge_sort', 'Stooge Sort algorithm'],
    'strandsort': ['Sorting_algorithms/Strand_sort', 'Strand Sort algorithm'],
    'subcipher': ['Substitution_cipher', 'Substitution Cipher'],
    'toposort': ['Topological_sort', 'Topological Sort Algorithm']
}


'''
    SUPPORTED_LANGUAGE configuration

    key: user input language flag
    value[0]:  Rosseta Code language format
    value[1]:  Formal language name to be displayed as output title
'''
SUPPORTED_LANGUAGES = {
    'actionscript': ['ActionScript', 'Actionscript'],
    'ada': ['Ada', 'Ada'],
    'algol68': ['ALGOL_68', 'ALGOL68'],
    'applescript': ['AppleScript', 'Applescript'],
    'autohotkey': ['AutoHotkey', 'Autohotkey'],
    'awk': ['AWK', 'AWK'],
    'c': ['C', 'C'],
    'cpp': ['C.2B.2B', 'C++'],
    'csharp': ['C.23', 'C#'],
    'd': ['D', 'D'],
    'delphi': ['Delphi', 'Delphi'],
    'fsharp': ['F.23', 'F#'],
    'eiffel': ['Eiffel', 'Eiffel'],
    'fortran': ['Fortran', 'Fortran'],
    'go': ['Go', 'Go'],
    'haskell': ['Haskell', 'Haskell'],
    'objc': ['Objective-C', 'Objective-C'],
    'java': ['Java', 'Java'],
    'javascript': ['JavaScript', 'Javascript'],
    'lua': ['Lua', 'Lua'],
    'matlab': ['MATLAB', 'Matlab'],
    'ocaml': ['Ocaml', 'Ocaml'],
    'pascal': ['Pascal', 'Pascal'],
    'perl': ['Perl', 'Perl'],
    'php': ['PHP', 'PHP'],
    'powershell': ['PowerShell', 'PowerShell'],
    'python': ['Python', 'Python'],
    'ruby': ['Ruby', 'Ruby'],
    'rust': ['Rust', 'Rust'],
    'scala': ['Scala', 'Scala'],
    'swift': ['Swift', 'Swift']

}

if __name__ == '__main__':
     print('This file is not intended to be called by the user. See algocli --help')
